### PR TITLE
statistics: implement FM Sketch

### PR DESCRIPTION
**Pull request by lamxTyler**
This PR implements the FM Sketch. For each sketch, we maintain it's hash set and current level.
Every time we insert a hashed value of the original value into the set:

* we first check if all the low level bits is 0, if not, we throw this value away.
* if so, we insert the value into set, then check if the hash set is full, if so,we increse current level, and remove all the value that has level-th bit set on.

When we merge sketches, we first set the result sketch's level to the max level among this sketches. Then repeatly insert the value in these sketches into the result sketch.

PTAL @coocood @hanfei1991 @zimulala @tiancaiamao @winoros

### CODE REVIEW

**Comment by shenli** [comment]
@lamxTyler Please add some description for the PR.

File: plan/statistics/sketch.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import (
+	"hash/fnv"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// Sketch is used to count the number of distinct elements in a set.
+type Sketch struct {
```
Comment (hanfei1991) [inline plan/statistics/sketch.go]: It's better to rename it as `FMSketch`. Sketch also has many kinds.

**Comment by lamxTyler** [comment]
This PR implemnts the FM Sketch. For each sketch, we maintain it's hash set and current level. 
Every time we insert a hash valued of the original row into the set:
* we first check if all the low `level` bits is 0, if not, we do throw this value away.
* if so, we then check if the hash set is full
     * if so,we increse current level, and remove all the value that has `level`-th bit set on.
     * otherwise, we insert the value into set.

When we merge sketches, we first set the result sketch's level to the max level among this sketches. Then repeatly insert the value in these sketches into the result sketch.

File: plan/statistics/sketch.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		}
+	} else {
+		s.hashset[hash] = true
+	}
+}
+
+func buildSketch(rows []types.Datum, maxSize int) (*Sketch, int64, error) {
+	s := newSketch()
+	h := fnv.New64a()
+	for _, row := range rows {
```
Comment (lamxTyler) [inline plan/statistics/sketch.go]: Each columns, here `rows` means all the rows of a single column.

File: statistics/fmsketch.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+func newFMSketch() *FMSketch {
+	return &FMSketch{hashset: make(map[uint64]bool)}
+}
+
+func (s *FMSketch) insertHashValue(hash uint64, maxSize int) {
+	mask := (1 << s.level) - uint64(1)
+	if (hash & mask) != 0 {
+		return
+	}
+	if len(s.hashset) == maxSize {
```
Comment (coocood) [inline statistics/fmsketch.go]: In this branch, the new hash value is ignored.
How about this:
```
s.hashset[hash] = true
if len(s.hashset) == maxSize {
   ...
}
```

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@coocood @hanfei1991 PTAL

File: statistics/fmsketch.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+type FMSketch struct {
+	hashset map[uint64]bool
+	level   uint
+}
+
+func newFMSketch() *FMSketch {
+	return &FMSketch{hashset: make(map[uint64]bool)}
+}
+
+func (s *FMSketch) insertHashValue(hash uint64, maxSize int) {
```
Comment (hanfei1991) [inline statistics/fmsketch.go]: I think it's more reasonable to make `maxSize` as fields.

File: statistics/fmsketch.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	mask := (1 << s.level) - uint64(1)
```
Comment (hanfei1991) [inline statistics/fmsketch.go]: We can record the mask

File: statistics/fmsketch.go (hunk 4 above)
Comment (hanfei1991) [inline statistics/fmsketch.go]: hash -> hashVal is clearer

File: statistics/fmsketch.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		if err != nil {
+			return nil, 0, errors.Trace(err)
+		}
+		s.insertHashValue(h.Sum64(), maxSize)
+	}
+	ndv := (1 << s.level) * int64(len(s.hashset))
+	return s, ndv, nil
+}
+
+func mergeFMSketches(sketches []*FMSketch, maxSize int) (*FMSketch, int64) {
```
Comment (hanfei1991) [inline statistics/fmsketch.go]: FMSketch.Merge ?

File: statistics/fmsketch.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	if (hash & mask) != 0 {
+		return
+	}
+	s.hashset[hash] = true
+	if len(s.hashset) > maxSize {
+		mask = mask*2 + 1
+		s.level++
+		for key := range s.hashset {
+			if (key & mask) != 0 {
+				delete(s.hashset, key)
```
Comment (hanfei1991) [inline statistics/fmsketch.go]: Is it safe in loop?

Comment (coocood) [inline statistics/fmsketch.go]: Yes, it is.

**Comment by coocood** [comment]
LGTM

File: statistics/fmsketch.go (hunk 6 above)
Comment (lamxTyler) [inline statistics/fmsketch.go]: I feel it is quite weird to do so. For example, when you consider how it is used: call `buildFMSketch` in a loop so we have a list of sketch, then directly call this function, is it weried to have to create an empty sketch or used one of the list as the left hand operator?

**Pull request merged**